src/SharedWeb/Utilities/RequestLoggingMiddleware.cs

```diff
@@ -0,0 +1,106 @@
+﻿using System.Collections;
+using Bit.Core.Settings;
+using Bit.Core.Utilities;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+
+#nullable enable
+
+namespace Bit.SharedWeb.Utilities;
+
+public sealed class RequestLoggingMiddleware
+{
+    private readonly RequestDelegate _next;
+    private readonly ILogger<RequestLoggingMiddleware> _logger;
+    private readonly GlobalSettings _globalSettings;
+
+    public RequestLoggingMiddleware(RequestDelegate next, ILogger<RequestLoggingMiddleware> logger, GlobalSettings globalSettings)
+    {
+        _next = next;
+        _logger = logger;
+        _globalSettings = globalSettings;
+    }
+
+    public Task Invoke(HttpContext context)
+    {
+        using (_logger.BeginScope(
+          new RequestLogScope(context.GetIpAddress(_globalSettings),
+            GetHeaderValue(context, "user-agent"),
+            GetHeaderValue(context, "device-type"),
+            GetHeaderValue(context, "device-type"))))
+        {
+            return _next(context);
+        }
+
+        static string? GetHeaderValue(HttpContext httpContext, string header)
```

Comment:
🌱 As an aside I could see this being a shared utility method. Tons of potential uses: https://github.com/search?q=repo%3Abitwarden%2Fserver%20httpContext.Request.Headers&type=code

Reply:
I'm not so sure, most of those uses want to do something only when there is a value. They'd likely be better served by using `TryGetValue` themselves, vs calling this helper and then having to do a null check on if we have back a value. 

I'd love for us to enable this rule as a warning: https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1854